File: GameLibrary/src/cc/lib/zombicide/ZArmor.java
package cc.lib.zombicide;

import cc.lib.game.Utils;
import cc.lib.utils.Table;

public class ZArmor extends ZEquipment<ZArmorType> {

    static {
        addAllFields(ZArmor.class);
    }

    public ZArmor() {
        this(null);
    }

    ZArmor(ZArmorType type) {
        this.type = type;
    }

    private final ZArmorType type;

    @Override
    public ZEquipSlotType getSlotType() {
        return type.slotType;
    }

    @Override
    public boolean canEquip() {
        return true;
    }

    @Override
    public boolean isArmor() {
        return true;
    }

    int getRating(ZZombieType type) {
        return this.type.getRating(type);
    }

    @Override
    public ZArmorType getType() {
        return type;
    }

    @Override
    public Table getCardInfo(ZCharacter c, ZGame game) {

        /*

        SHIELD
        specialInfo    |
        ---------------| How armor works text
        RATING         |
        Walker      | 0|
        Fatty       | 5|
        Runner      | 5|
        Necromancer | 5|
        Abomination | 5|
        Special        |
         */



        Table ratings = new Table().setNoBorder();
        for (ZZombieType type : Utils.asList(ZZombieType.Walker, ZZombieType.Fatty, ZZombieType.Runner, ZZombieType.Necromancer, ZZombieType.Abomination)) {
            ratings.addRow(type, getRating(type));
        }
        Table main = new Table(getLabel()).setNoBorder().addRow(ratings);
        if (type.specialAbilityDescription != null) {
            main.addRow(Utils.wrapTextWithNewlines(type.specialAbilityDescription, 24));
        }
        return main;
    }

    @Override
    public String getTooltipText() {
        Table ratings = new Table().setNoBorder();
        for (ZZombieType type : Utils.asList(ZZombieType.Walker, ZZombieType.Fatty, ZZombieType.Runner, ZZombieType.Necromancer, ZZombieType.Abomination)) {
            ratings.addRow(type, getRating(type));
        }
        if (type.specialAbilityDescription != null) {
            ratings.addRow(Utils.wrapTextWithNewlines(type.specialAbilityDescription, 24));
        }
        return ratings.toString();
    }
}

File: GameLibrary/src/cc/lib/zombicide/ZArmor.kt
package cc.lib.zombicide

import cc.lib.game.Utils
import cc.lib.utils.Table

class ZArmor(override val type: ZArmorType=ZArmorType.CHAINMAIL) : ZEquipment<ZArmorType>() {
    companion object {
        init {
            addAllFields(ZArmor::class.java)
        }
    }

    override val slotType: ZEquipSlotType
        get() = type.slotType

    override fun isEquippable(c: ZCharacter): Boolean {
        return true
    }

    override val isArmor: Boolean
        get() = true

    override fun getCardInfo(c: ZCharacter, game: ZGame): Table {

        /*

        SHIELD
        specialInfo    |
        ---------------| How armor works text
        RATING         |
        Walker      | 0|
        Fatty       | 5|
        Runner      | 5|
        Necromancer | 5|
        Abomination | 5|
        Special        |
         */
        val ratings = Table().setNoBorder()
        for (type in Utils.asList(ZZombieType.Walker, ZZombieType.Fatty, ZZombieType.Runner, ZZombieType.Necromancer, ZZombieType.Abomination)) {
            ratings.addRow(type, this.type.getDieRollToBlock(type))
        }
        val main = Table(label).setNoBorder().addRow(ratings)
        if (type.specialAbilityDescription != null) {
            main.addRow(Utils.wrapTextWithNewlines(type.specialAbilityDescription, 24))
        }
        return main
    }

    override fun getTooltipText(): String {
        val ratings = Table().setNoBorder()
        for (type in Utils.asList(ZZombieType.Walker, ZZombieType.Fatty, ZZombieType.Runner, ZZombieType.Necromancer, ZZombieType.Abomination)) {
            ratings.addRow(type, this.type.getDieRollToBlock(type))
        }
        if (type.specialAbilityDescription != null) {
            ratings.addRow(Utils.wrapTextWithNewlines(type.specialAbilityDescription, 24))
        }
        return ratings.toString()
    }
}